code/network/multimsgs.cpp

```diff
@@ -2161,79 +2161,218 @@ void process_netgame_update_packet( ubyte *data, header *hinfo )
 //*********************************************************************************************************
 // Game Update Packet
 //*********************************************************************************************************
-/*
-
-*/
+//
 // send a request or a reply for mission description, if code == 0, request, if code == 1, reply
-void send_netgame_descript_packet(net_addr *addr, int code)
+// Cyborg17 - Now recursive for longer descriptions! Original call needs to check if this is a campaign first. 
+// Message order is Campaign desc 1, Opt. Campaign addendum, Mission desc 1, then Opt. mission desc addendum
+// Single missions will skip straight to Mission desc 1.
+void send_netgame_descript_packet(net_addr *addr, int code, bool campaign, bool addendum, ubyte message_count)
 {
-	ubyte data[MAX_PACKET_SIZE],val;
-	int desc_len;
+	// Get this out of the way, because we don't want to waste our time if the addr is bad.
+	Assert(addr != nullptr);
+	if (addr == nullptr) {
+		return;	
+	}
+	
+	ubyte data[MAX_PACKET_SIZE];
+	ubyte packet_type = 0;
 	int packet_size = 0;
-
+	
 	// build the header
 	BUILD_HEADER(UPDATE_DESCRIPT);
+	
+	if (code) {
 
-	val = (ubyte)code;
-	ADD_DATA(val);	
+		int desc_len = 0;
+		bool second_packet_needed = false;
+		char message[MAX_PACKET_DESC_LEN], temp[2 * MAX_PACKET_DESC_LEN];
+		memset(message, 0, MAX_PACKET_DESC_LEN);
+		memset(temp, 0, 2 * MAX_PACKET_DESC_LEN);
+		
+		if (campaign) {
+			// bitwise assignment to save bandwidth
+			memcpy(temp, Netgame.campaign_desc, 2 * MAX_PACKET_DESC_LEN);
+
+			// check to see if the string is too long for one packet
+			desc_len = (int)strlen(temp);
+
+			// No actual campaign description
+			if (desc_len <= 0) {
+				strcpy_s(message, PRE_CAMPAIGN_DESC);
+				strcat_s(message, "No campaign description.");
+				packet_type |= DESCRIPT_PACKET_CAMPAIGN_MESSAGE_1;
+			}			// subtract because of the text that marks this as a campaign description.
+			else if (!addendum && desc_len > (MAX_PACKET_DESC_LEN - 24)) {
+				// if we are going to need to recurse....
+				second_packet_needed = true;
+				desc_len = MAX_PACKET_DESC_LEN;
+				strcpy_s(message, PRE_CAMPAIGN_DESC);
+				strncat(message, temp, MAX_PACKET_DESC_LEN - 23);
+				packet_type |= DESCRIPT_PACKET_CAMPAIGN_MESSAGE_1;
+
+			} else if (addendum) {
+				// if we now recursing
+				backspace(temp, MAX_PACKET_DESC_LEN - 24); // Subtract because we had the campaign desc intro text.
+				desc_len -= (MAX_PACKET_DESC_LEN - 24);
```

Comment:
What's with all of the magic numbers? 23, 24, 27, 28? If they are needed then reference them with some #define's so that future coders won't need to figure out what they're for.

Reply:
They were just the length of the constant strings, I changed it to strlen()